pkg/golinters/revive.go

```diff
@@ -0,0 +1,187 @@
+
+	"github.com/golangci/golangci-lint/pkg/config"
+	"github.com/golangci/golangci-lint/pkg/golinters/goanalysis"
+	"github.com/golangci/golangci-lint/pkg/lint/linter"
+	"github.com/golangci/golangci-lint/pkg/result"
+
+	reviveConfig "github.com/mgechev/revive/config"
+	"github.com/mgechev/revive/lint"
+	"golang.org/x/tools/go/analysis"
+)
+
+const (
+	reviveName = "revive"
+)
+
+// jsonObject defines a JSON object of an failure
+type jsonObject struct {
+	Severity     lint.Severity
+	lint.Failure `json:",inline"`
+}
+
+// NewNewRevive returns a new Revive linter.
+func NewRevive(cfg *config.ReviveSettings) *goanalysis.Linter {
+	var (
+		issues   []goanalysis.Issue
+		analyzer = &analysis.Analyzer{
+			Name: goanalysis.TheOnlyAnalyzerName,
+			Doc:  goanalysis.TheOnlyanalyzerDoc,
+		}
+	)
+
+	return goanalysis.NewLinter(
+		reviveName,
+		"Fast, configurable, extensible, flexible, and beautiful linter for Go",
+		[]*analysis.Analyzer{analyzer},
+		nil,
+	).WithContextSetter(func(lintCtx *linter.Context) {
+		analyzer.Run = func(pass *analysis.Pass) (interface{}, error) {
+			var (
+				files = []string{}
+			)
+
+			for _, file := range pass.Files {
+				files = append(files, pass.Fset.PositionFor(file.Pos(), false).Filename)
+			}
+
+			conf, err := SetReviveConfig(cfg)
+			if err != nil {
+				return nil, err
+			}
+
+			formatter, err := reviveConfig.GetFormatter("json")
+			if err != nil {
+				return nil, err
+			}
+
+			revive := lint.New(ioutil.ReadFile)
+
+			lintingRules, err := reviveConfig.GetLintingRules(conf)
+			if err != nil {
+				return nil, err
+			}
+
+			packages, err := dots.ResolvePackages(files, normalizeSplit([]string{}))
+			if err != nil {
+				return nil, err
+			}
+
+			failures, err := revive.Lint(packages, lintingRules, *conf)
+			if err != nil {
+				return nil, err
+			}
+
+			formatChan := make(chan lint.Failure)
+			exitChan := make(chan bool)
+
+			var output string
+			go (func() {
+				output, _ = formatter.Format(formatChan, *conf)
+				exitChan <- true
+			})()
+
+			for f := range failures {
+				if f.Confidence < conf.Confidence {
+					continue
+				}
+
+				formatChan <- f
+			}
+
+			close(formatChan)
+			<-exitChan
+
+			var results []jsonObject
+			err = json.Unmarshal([]byte(output), &results)
+			if err != nil {
+				return nil, err
+			}
+
+			for i := range results {
+				issues = append(issues, goanalysis.NewIssue(&result.Issue{
+					Severity: string(results[i].Severity),
+					Text:     fmt.Sprintf("%q", results[i].Failure.Failure),
+					Pos: token.Position{
+						Filename: results[i].Position.Start.Filename,
+						Line:     results[i].Position.Start.Line,
+						Offset:   results[i].Position.Start.Offset,
+						Column:   results[i].Position.Start.Column,
+					},
+					LineRange: &result.Range{
+						From: results[i].Position.Start.Line,
+						To:   results[i].Position.End.Line,
+					},
+					FromLinter: reviveName,
+				}, pass))
+			}
+			return nil, nil
+		}
```

Comment:
```suggestion
			for _, r := range results {
				issues = append(issues, goanalysis.NewIssue(&result.Issue{
					Severity: string(r.Severity),
					Text:     fmt.Sprintf("%q", r.Failure.Failure),
					Pos: token.Position{
						Filename: r.Position.Start.Filename,
						Line:     r.Position.Start.Line,
						Offset:   r.Position.Start.Offset,
						Column:   r.Position.Start.Column,
					},
					LineRange: &result.Range{
						From: r.Position.Start.Line,
						To:   r.Position.End.Line,
					},
					FromLinter: reviveName,
				}, pass))
			}
```

Reply:
I agree the suggested change looks nicer, but gocritic complains about it and suggests pointers or indexing which causes the build to fail.
```
pkg\golinters\revive.go:108:4: rangeValCopy: each iteration copies 184 bytes (consider pointers or indexing) (gocritic)
                        for _, r := range results {
                        ^
```

Reply:
the other suggestion a more important, so keep your current code on that point.